refactor(compiler): support interpolation and encoded entities when lexing markup

petebacondarwin opened the pull request:
The lexer now splits interpolation and encoded entity tokens out from the string it is tokenizing.

See individual commits...

Fixes #41034

petebacondarwin committed:
refactor(compiler): support interpolation tokens when lexing markup

The lexer now splits interpolation tokens out from text tokens.

Previously the contents of `<div>Hello, {{ name}}<div>` would be a single
text token. Now it will be three tokens:

```
TEXT: "Hello, "
INTERPOLATION: "{{", " name", "}}"
TEXT: ""
```

- INTERPOLATION tokens have three parts, "start marker", "expression"
  and "end marker".
- INTERPOLATION tokens are always preceded and followed by TEXT tokens,
  even if they represent an empty string.

The HTML parser has been modified to recombine these tokens to allow this
refactoring to have limited effect in this commit. Further refactorings
to use these new tokens will follow in subsequent commits.

petebacondarwin committed:
refactor(compiler): share `isQuote()` via `chars.ts`

This function is general purpose and by moving it into the
`chars.ts` file along with similar helpers, it can be reused
in the lexer, for instance.

petebacondarwin committed:
refactor(compiler): support interpolation tokens when lexing attribute values

The lexer now splits interpolation tokens out from attribute value tokens.
Previously the attribute value of `<div attr="Hello, {{ name}}">` would be a single
token. Now it will be three tokens:

```
ATTR_VALUE_TEXT: "Hello, "
ATTR_VALUE_INTERPOLATION: "{{", " name", "}}"
ATTR_VALUE_TEXT: ""
```

- ATTR_VALUE_INTERPOLATION tokens have three parts, "start marker",
  "expression" and "end marker".
- ATTR_VALUE_INTERPOLATION tokens are always preceded and followed
  by TEXT tokens, even if they represent an empty string.

The HTML parser has been modified to recombine these tokens to allow this
refactoring to have limited effect in this commit. Further refactorings
to use these new tokens will follow in subsequent commits.

petebacondarwin committed:
refactor(compiler): support encoded entity tokens when lexing markup

The lexer now splits encoded entity tokens out from text and attribute value tokens.

Previously encoded entities would be decoded and the decoded value would be
included as part of the text token of the surrounding text. Now the entities
have their own tokens. There are two scenarios: text and attribute values.

Previously the contents of `<div>Hello &amp; goodbye</div>` would be a single
TEXT token. Now it will be three tokens:

```
TEXT: "Hello "
ENCODED_ENTITY: "&", "&amp;"
TEXT: " goodbye"
```

Previously the attribute value in `<div title="Hello &amp; goodbye">` would be
a single text token. Now it will be three tokens:

```
ATTR_VALUE_TEXT: "Hello "
ENCODED_ENTITY: "&", "&amp;"
ATTR_VALUE_TEXT: " goodbye"
```

- ENCODED_ENTITY tokens have two parts: "decoded" and "encoded".
- ENCODED_ENTITY tokens are always preceded and followed by either TEXT tokens
  or ATTR_VALUE_TEXT tokens, depending upon the context, even if they represent
  an empty string.

The HTML parser has been modified to recombine these tokens to allow this
refactoring to have limited effect in this commit. Further refactorings
to use these new tokens will follow in subsequent commits.

petebacondarwin committed:
refactor(compiler): expose token parts in Text nodes

When it was tokenized, text content is split into parts that can include
interpolations and encoded entities tokens.

To make this information available to downstream processing, this commit
adds these tokens to the `Text` AST nodes, with suitable processing.

petebacondarwin committed:
refactor(compiler): remove cyclic dependencies

This commit removes 9 cycles in the dependency graph of the compiler code.

petebacondarwin committed:
test(compiler-cli): clarify source-map expectations in compliance tests

The compliance tests can check source-map segments against expectations
encoded into the expectation files. Previously, the encoding of the expected
segment was only delimited by whitespace, but this made it difficult to identify
segments that started or ended with whitespace.

Now these segment expectations are wrapped in double-quotes which makes
it easier to read and understand the expectation files.

petebacondarwin committed:
test(compiler): check `fullStart` source-span

The tests were checking that the source-span of parsed HTML nodes were
accurate, but they were not checking the span when it includes the
"leading trivia", which are given by the `fullStart` rather than `start`
location.

petebacondarwin committed:
fix(compiler): include leading whitespace in source-spans of i18n messages

Previously, the way templates were tokenized meant that we lost information
about the location of interpolations if the template contained encoded HTML
entities. This meant that the mapping back to the source interpolated strings
could be offset incorrectly.

Also, the source-span assigned to an i18n message did not include leading
whitespace. This confused the output source-mappings so that the first text
nodes of the message stopped at the first non-whitespace character.

This commit makes use of the previous refactorings, where more fine grain
information was provided in text tokens, to enable the parser to identify
the location of the interpolations in the original source more accurately.

Fixes #41034

jelbourn reviewed (approved):
LGTM

Reviewed-for: fw-playground

JoostK commented on packages/compiler/src/ml_parser/lexer.ts:
```diff
@@ -435,21 +433,24 @@ class _Tokenizer {
       this._attemptCharCodeUntilFn(isNamedEntityEnd);
       if (this._cursor.peek() != chars.$SEMICOLON) {
+        // No semicolon was found so treat this as text
+        this._beginToken(textTokenType, start);
```
I got a bit confused here because this appears to leave the `this._beginToken(TokenType.ENCODED_ENTITY);` at the method's start unclosed, but looking at it more closely that doesn't matter. Maybe it would be clearer to only begin the `ENCODED_ENTITY` token if we are sure it is parsed as such.

petebacondarwin replied:
That would require calling beginToken in multiple places for the main use cases of this function.
I feel that since this is an alternative use case, it is better to make this the exception. I'll add a comment to make this more clear.

JoostK commented on packages/compiler/src/ml_parser/lexer.ts:
```diff
@@ -866,7 +879,9 @@ function mergeTextTokens(srcTokens: Token[]): Token[] {
   for (let i = 0; i < srcTokens.length; i++) {
     const token = srcTokens[i];
-    if (lastDstToken && lastDstToken.type == TokenType.TEXT && token.type == TokenType.TEXT) {
+    if ((lastDstToken && lastDstToken.type == TokenType.TEXT && token.type == TokenType.TEXT) ||
```
Since you're touching this line consider changing the `==` to `===`.

petebacondarwin replied:
There are `==` all over this file. I'll refactor the lot in a separate commit.

JoostK commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -94,6 +99,14 @@ export class WhitespaceVisitor implements html.Visitor {
 }
 
+function processTokenWhitespace(token: Token): Token {
+  return {
```
`Token` is a class so this should use `new Token`

petebacondarwin replied:
In the very last commit, I now refactor the whole `Token` stuff, so that it is now an interface, not a class. PTAL.

JoostK commented on packages/compiler/src/ml_parser/html_whitespaces.ts:
```diff
@@ -94,6 +99,14 @@ export class WhitespaceVisitor implements html.Visitor {
+  return {
+    type: token.type,
+    sourceSpan: token.sourceSpan,
+    parts: [token.parts[0].replace(WS_REPLACE_REGEXP, ' ')]
```
It's kind of magical that this only selects the first part and rejects any remaining ones.

Also, would this one need `replaceNgsp` here, as it is used like that in `visitText`?

petebacondarwin replied:
At this point the `Text` tokens should only have one `part`.

petebacondarwin replied:
The last commit refactors Tokens so that it is clearer what the parts are for different TokenTypes

JoostK commented on packages/compiler/src/ml_parser/parser.ts:
```diff
@@ -215,21 +215,27 @@ class _TreeBuilder {
+        tokens[0] = {
+          type: token.type,
+          sourceSpan: token.sourceSpan,
+          parts: [token.parts[0].substring(1)]
```
Only keeping the first part here is also quite magical.

petebacondarwin replied:
Actually we can just use `text` here.
(And text tokens only have one `part` so this is why we don't iterate over all parts.)

JoostK commented on packages/compiler/src/i18n/i18n_parser.ts:
```diff
@@ -165,66 +171,37 @@ class _I18nVisitor implements html.Visitor {
-    for (let i = 0; i < strings.length - 1; i++) {
-      this._addText(nodes, strings[i], sourceSpan);
-      this._addPlaceholder(nodes, context, expressions[i], sourceSpan);
+    for (let i = 0; i < tokens.length; i++) {
```
Looks like this can become a for-of loop now that `i` is no longer used to index into two distinct arrays.

JoostK commented on packages/compiler/src/i18n/i18n_parser.ts:
```diff
@@ -165,66 +171,37 @@ class _I18nVisitor implements html.Visitor {
+      switch (token.type) {
+        case TokenType.INTERPOLATION:
+        case TokenType.ATTR_VALUE_INTERPOLATION:
+          const expression = token.parts[1];
```
I wish the token parts would be documented better, this is a bit magical. Perhaps a const enum would help clarify these indices.

petebacondarwin replied:
I am going to add discriminated interfaces for each token type in a separate commit.

petebacondarwin replied:
This has been done in the last commit.

JoostK reviewed (commented):
I left a bunch of comments. I would be curious to know how this would behave in TGP run, as that would give us a better feeling for how breaky this might be and/or which cases this doesn't handle correctly.

---

There's two typos in "refactor(compiler): support encoded entity tokens when lexing markup":

> The lexer now splits encoded entity tokens out from text and attribute value tokens.
Previously encoded entities would be decoded and **~there~ their** decoded value would be
included as part of the text token of the surrounding text. Now the entities will have
their own tokens. There are two scenarios: text and attribute values.

> The HTML parser has been modified to **~recombined~ recombine** these tokens to allow this
refactoring to have limited effect in this commit. Further refactorings
to use these new tokens will follow in subsequent commits.

petebacondarwin commented on modules/playground/src/zippy_component/app/zippy.html:
```diff
@@ -1,6 +1,6 @@
 <div class="zippy">
   <div (click)="toggle()" class="zippy__title">
-    {{ visible ? '&#x25BE;' : '&#x25B8;' }} {{title}}
+    {{ visible ? '\u25BE' : '\u25B8' }} {{title}}
```
One person's bug fix is another person's breaking change... but perhaps you are right 😞 

No doubt there will be loads of these in G3...

petebacondarwin replied:
Hmm, this was a refactoring commit. I wonder how I can get this BREAKING CHANGE into the change log...
Perhaps I should rename this commit to a `fix`? Where I am "fixing" the fact that interpolated blocks should not be decoding HTML entities??

petebacondarwin replied:
@alxhub @JoostK - thoughts?

petebacondarwin replied:
I have added a commit that ensures this breaking change does not happen.

petebacondarwin commented on packages/compiler/src/ml_parser/lexer.ts:
```diff
@@ -716,6 +714,54 @@ class _Tokenizer {
   }
 
+  private _consumeInterpolation(interpolationStart: CharacterCursor) {
+    const parts: string[] = [];
```
I think that carriage returns in interpolation expressions are not so common, but I will add that processing to the individual parts.

petebacondarwin replied:
We don't want to join these but I have added in the CR processing for each part..

alxhub commented on packages/compiler/src/ml_parser/ast.ts:
```diff
@@ -23,7 +24,8 @@ export abstract class NodeWithI18n implements BaseNode {
 export class Text extends NodeWithI18n {
-  constructor(public value: string, sourceSpan: ParseSourceSpan, i18n?: I18nMeta) {
+  constructor(
+      public value: string, sourceSpan: ParseSourceSpan, public tokens: Token[], i18n?: I18nMeta) {
```
I think we should create a separate AST type for each `Token` that can appear here, instead of exposing the lexer type in the final AST. There are also many token variants that can't appear in `tokens` here, so narrowing to a union of only a few different kinds would definitely be wise.

We could still use the same runtime _value_ for the tokens, as long as the new interface was structurally compatible with `Token`.

petebacondarwin replied:
This fits with @JoostK's request to make the `parts` more clear via interfaces.
I'll add a refactoring commit to the beginning of this PR.

petebacondarwin replied:
Such a refactoring has been added to the "end" of the PR. (Putting it at the beginning caused too much churn in all the other changes.)

petebacondarwin commented on packages/compiler/src/ml_parser/lexer.ts:
```diff
@@ -716,6 +714,54 @@ class _Tokenizer {
+        if (chars.isAsciiLetter(this._cursor.peek())) {
+          // We are starting what looks like an HTML element in the middle of this interpolation.
+          // Reset the cursor to before the `<` character and end the interpolation token.
+          // (This is actually wrong but here for backward compatibility).
```
There is no end delimiter in the template, so adding one would be disingenuous IMO.
By leaving that part `undefined` downstream processors can choose to do something different.

petebacondarwin commented on packages/compiler/src/ml_parser/lexer.ts:
```diff
@@ -716,6 +714,54 @@ class _Tokenizer {
+        return this._endToken(parts);
+      }
+
+      const char = this._readChar(true);
```
Very good call!

petebacondarwin replied:
I've added handling for this.

petebacondarwin commented on packages/compiler/src/ml_parser/lexer.ts:
```diff
@@ -716,6 +714,54 @@ class _Tokenizer {
+      } else if (char === inQuote) {
+        // Exiting the current quoted string
+        inQuote = null;
+      } else if (/['"`]/.test(char)) {
```
> Also, I don't think backticks (template strings) are legal quotes in Angular templates.

Yes, I didn't realise that!

petebacondarwin replied:
Later on this code reuses the `isQuote()` helper from the expression parser, which does also lex backticks. I think it is best if we keep them inline in the lexing, and then throw errors in the parsing if they are not supported.

petebacondarwin committed:
refactor(compiler): define interfaces for each lexer token

These token interfaces will make it easier to reason about tokens in the
parser and in specs.

Previously, it was never clear what items could appear in the `parts`
array of a token given a particular `TokenType`. Now, each token interface
declares a labelled tuple for the parts, which helps to document the token
better.

petebacondarwin commented on packages/compiler/src/ml_parser/parser.ts:
```diff
@@ -349,27 +349,41 @@ class _TreeBuilder {
+      // For now we are recombining text and interpolation tokens
+      while (this._peek.type === lex.TokenType.ATTR_VALUE_TEXT ||
+             this._peek.type === lex.TokenType.ATTR_VALUE_INTERPOLATION) {
+        let valueToken = this._advance();
```
👍

AndrewKushnir commented:
[Presubmit](http://test/OCL:379395451:BASE:379395459:1623722371211:41b5e1e0).

AndrewKushnir commented:
Quick updated after running a subset of targets in g3 (blueprint presubmit): there are few broken targets, related to the changes in this PR. I'll investigate more and try to extract specific cases (in a form of a minimal repro), so that we can also extend our test coverage.

AndrewKushnir commented:
FYI I've started [a new presubmit](http://test/OCL:379395451:BASE:379579995:1623793954177:de316e82) with the most recent changes from this PR, since there were some changes that would help resolve issues in the failing targets (thanks @JoostK for bringing it up!). Will keep this thread updated.

AndrewKushnir commented:
Another update after running a new presubmit: it looks like the previous problem went away. There are few broken targets currently caused by unterminated interpolation expressions in a template (note the missing `}}`):
```
<my-comp title="{{ a ? 'a' : 'b' "> ... </my-comp>
```

The error message is:

```
error TS-995002: Opening tag "my-comp" not terminated.
```

I will update templates as needed, but wanted to share this in case there is an easy way to detect this situation and throw an error about unterminated interpolation vs not terminated tag (and to check whether this makes sense or the current error is better).

petebacondarwin commented:
Thanks @AndrewKushnir, I think I can make the error message better in these cases.

AndrewKushnir commented:
One more case that I was able to extract from the broken targets:
```
<section>
  <code>
    &lt;ng-template let-suggestion&gt;
      {{suggestion}<!---->}
    &lt;/ng-template&gt;
  </code>
  <section></section>
</section>
```
the error message is:
```
error TS-995002: Unexpected closing tag "section". It may happen when the tag has already been closed by another tag. For more info see https://www.w3.org/TR/html5/syntax.html#closing-elements-that-have-implied-end-tags

8 </section>
  ~~~~~~~~~~
```
It looks like everything after `{{` is considered a content of an interpolation (which might be correct?). A couple additional things I tried:

- removing the inner `<section></section>` avoids the problem (which is a bit surprising)
- adding `ngNonBindable` to the `<code>` element doesn't help (seems to have no effect), which might be something we'd need to investigate

@petebacondarwin please let me know if you're able to repro this locally as well (just tested that in g3, not locally).

petebacondarwin commented:
I think what is happening in that case is that it is being read as:

```html
<section>
  <code>
    &lt;ng-template let-suggestion&gt;
    {{suggestion}<!---->}&lt;/ng-template&gt;</code><section>
</section>
</section>
```

petebacondarwin committed:
test(compiler): check that the parser supports prematurely terminated interpolations

Such interpolations turned up during internal testing at Google, so this
commit adds a test to prevent regressions.

petebacondarwin commented:
@AndrewKushnir - I think I have removed and potential new error reports, to ensure backward compatibility. Would you be able to start a new presubmit perhaps?

AndrewKushnir commented:
[Presubmit](http://test/OCL:379395451:BASE:380044962:1623965052533:439a9dc6).

AndrewKushnir commented:
New presubmit indicated a different breakage now. Here is a minimal repro:
```
<pre>import {A} from 'a';</pre>
```
The error message:
```
error TS-995002: Unexpected character "EOF" (Do you have an unescaped "{" in your template? Use "{{ '{' }}") to escape it.)
```
While the error makes sense, it seems like it used to work before.

petebacondarwin commented:
I have found that this is actually a bug in the original lexer and created a pre-emptive fix for it, which we should land before this PR: https://github.com/angular/angular/pull/42605

petebacondarwin commented:
I've rebased this on top of #42605 so that we can get some feedback from CI. Once that other PR lands, I will rebase again and then ask for more reviews.

alxhub commented on packages/compiler-cli/ngcc/test/execution/helpers.ts:
```diff
@@ -6,7 +6,7 @@
  */
 import {DepGraph} from 'dependency-graph';
-import {DtsProcessing, PartiallyOrderedTasks, Task} from '../../src/execution/tasks/api';
+import {DtsProcessing, PartiallyOrderedTasks} from '../../src/execution/tasks/api';
```
This seems like a spurious change - should it be included here?

petebacondarwin replied:
Very spurious! Not sure how that ended up in this PR!

Since this `Task` is not used in the file, it is a valid clean up. I'll move it to its own commit in this PR, since it would be a bit overkill to create a whole PR just for this.

alxhub commented on packages/compiler/src/ml_parser/tokens.ts:
```diff
@@ -0,0 +1,172 @@
+    AttributeValueTextToken|AttributeValueInterpolationToken|EncodedEntityToken;
+
+export interface TokenBase {
+  type: TokenType|null;
```
+1 - yay for catching programming errors with types!

petebacondarwin commented on packages/compiler/src/ml_parser/lexer.ts:
```diff
@@ -720,6 +718,61 @@ class _Tokenizer {
+          return this._endToken(parts);
+        } else if (this._attemptStr('//')) {
+          // Once we are in a comment we ignore any quotes
+          inComment = true;
```
Was that a test with the current compiler, or after this PR?

Are you suggesting that we leave this comment here because something should change in this PR? Or just as a reminder that we might want to revisit this in a future PR?

petebacondarwin committed:
refactor(ngcc): remove unused import

This import is not used in the file, so can be removed.

petebacondarwin committed:
refactor(compiler): use `===` rather than `==` in the ml_parser

This is a simple tidy up commit to move to the more specific `===`
comparison operator in the HTML lexer/parser.

AndrewKushnir commented:
[Presubmit](http://test/OCL:384250765:BASE:384250948:1626107178061:c834ba50) (note: we'll also need to start a global one).

AndrewKushnir commented:
@petebacondarwin quick update:
- there is 1 failing target (likely caused by these changes) in the regular presubmit. I'll investigate it more this week and try to come up with a local repro.
- I was unable to start a global presubmit due to the merge conflicts :( Please let me know once the PR is rebased and I will try to run a global presubmit again.

AndrewKushnir commented:
[Global Presubmit](https://test.corp.google.com/OCL:385334299:BASE:385334350:1626546947811:b4b394fb).

petebacondarwin committed:
test(compiler): add a test for parsing multiline expressions in attributes

This tests a scenario that was failing in an internal project.

AndrewKushnir commented:
[Global Presubmit #2](https://test.corp.google.com/OCL:386157821:BASE:386157840:1626927475889:9a0fc7b0).

AndrewKushnir commented:
[Global Presubmit #3](https://test.corp.google.com/OCL:388142618:BASE:388142664:1627878964854:98434b4e).

AndrewKushnir commented:
FYI, global presubmit is successful, this PR should be ready for merge!

AndrewKushnir commented:
FYI, reopening this PR since it was rolled back due to the issues in g3. Specifically, extracted message (xi18n) that contained `&ngsp;`, `&amp;` (and potentially other HTML entities) had a different message id, triggering re-translations. Unfortunately, these problems were not caught by the global presubmit (TGP). We would need to make sure that:

- we have extra test coverage for xi18n for messages with HTML entities
- there are extra presubmits internally that verify that no message ids were changed unexpectedly

Also before landing the change again, we'd need to run i18n extraction for the affected app.

@petebacondarwin could you please have a look at the i18n extraction once you're back?

petebacondarwin commented:
Previously if there are interpolations in a message then you end up with an i18n “container” node, which is serialized to something like:

```
[text, ${ph}, text] 
```

After this PR, encoded entities are sort of treated like interpolations so we ended up getting:

```
'[before , &,  after]'
```

rather than

```
'before & after`
```

The digest value that is then generated from these strings is obviously different.
